lib/audio-device.js

```diff
@@ -0,0 +1,218 @@
+  }
+
+  toJSON() {
+    return {
+      format: this.format,
+      channels: this.channels,
+      freq: this.freq
+    }
+  }
+}
+
+class SDLAudioDeviceFormat {
+  constructor(deviceId) {
+    this._handle = binding.getAudioDeviceFormat(deviceId)
+  }
+
+  get valid() {
+    return binding.getAudioDeviceFormatValid(this._handle)
+  }
+
+  get sampleFrames() {
+    return binding.getAudioDeviceFormatSampleFrames(this._handle)
+  }
+
+  get spec() {
+    return new SDLAudioSpec(this)
+  }
+
+  toJSON() {
+    return {
+      valid: this.valid,
+      sampleFrames: this.sampleFrames,
+      spec: this.spec.toJSON()
+    }
+  }
+}
+
+class SDLAudioDevice extends ReadyResource {
+  static AudioSpec = SDLAudioSpec
+  static AudioDeviceFormat = SDLAudioDeviceFormat
+
+  /**
+   * @return []SDLAudioDeviceFormat
+   */
+  static playbackDeviceFormats() {
+    const list = binding.getAudioPlaybackDevices()
+    return list.map((deviceId) => {
+      return new SDLAudioDeviceFormat(deviceId)
+    })
+  }
+
+  /**
+   * @return []SDLAudioDeviceFormat
+   */
+  static recordingDeviceFormats() {
+    const list = binding.getAudioRecordingDevices()
+    return list.map((deviceId) => {
+      return new SDLAudioDeviceFormat(deviceId)
+    })
+  }
+
+  /**
+   * @return []SDLAudioDevice
+   */
+  static async recordingDevices() {
+    const list = binding.getAudioRecordingDevices()
+    return Promise.all(
+      list.map(async (deviceId) => {
+        const device = await SDLAudioDevice.open(deviceId)
+        return device
+      })
+    )
+  }
+
+  /**
+   * @return []SDLAudioDevice
+   */
+  static async playbackDevices() {
+    const list = binding.getAudioPlaybackDevices()
+    return Promise.all(
+      list.map(async (deviceId) => {
+        return SDLAudioDevice.open(deviceId)
+      })
+    )
+  }
+
+  /**
+   * @return SDLAudioDevice
+   */
+  static async defaultRecordingDevice(spec) {
+    return SDLAudioDevice.open(
+      constants.SDL_AUDIO_DEVICE_DEFAULT_PLAYBACK,
+      spec
+    )
+  }
+
+  /**
+   * @return SDLAudioDevice
+   */
+  static async defaultPlaybackDevice(spec) {
+    return SDLAudioDevice.open(
+      constants.SDL_AUDIO_DEVICE_DEFAULT_PLAYBACK,
+      spec
+    )
+  }
+
+  /**
+   * @return SDLAudioDevice
+   */
+  static async open(deviceId, spec) {
+    const device = new SDLAudioDevice(deviceId, spec)
+    await device.ready()
+    return device
+  }
+
+  constructor(deviceId, spec) {
+    super()
+    this._requestedDeviceId = deviceId
+    this.spec = spec
+    this.id = null
+  }
+
+  _open() {
+    if (this.opening || this.opened) {
+      return
+    }
+
+    const format = this.spec?.format
+    const channels = this.spec?.channels
+    const freq = this.spec?.freq
+
+    this.id = binding.openAudioDevice(
+      this._requestedDeviceId,
+      format,
+      channels,
+      freq
+    )
+  }
+
+  _close() {
+    if (this.closing) return
+    binding.closeAudioDevice(this.id)
+    this.id = null
+  }
```

Comment:
Move the `_open()` steps to the constructor and rename `_close()` to `destroy()` to match the existing APIs. We don't need `ReadyResource` for this as all of the resource management is fully synchronous. What we should do though is add `Symbol.dipose` methods so we can start using `using` like in `bare-ffmpeg`.

Reply:
Ah geez I was too ready to use ready-resource. Fixed!